nutils/evaluable.py

```diff
@@ -2375,8 +2375,8 @@ def __len__(self):
     return 2
 
   def __iter__(self):
-    yield ArrayFromTuple(self, index=0, shape=self.func.shape[:-1], dtype=complex if not self.symmetric else float)
-    yield ArrayFromTuple(self, index=1, shape=self.func.shape, dtype=complex if not self.symmetric or self.func.dtype == complex else float)
+    yield ArrayFromTuple(self, index=0, shape=self.func.shape[:-1], dtype=float if self.symmetric else complex)
+    yield ArrayFromTuple(self, index=1, shape=self.func.shape, dtype=self.func.dtype if self.symmetric else complex)
```

Comment:
I think we do want to keep the `dtype` rules here, but sometimes the `dtype` won't agree with underlying array's, do we only check `dtype` is 'safely' castable to the one in the object? (see also the 'cast' commit)
Note here the semantic is not changed, it's merely code simplification.

Reply:
Even if the matrix has real eigenvalues they will be returned as complex data, just with a zero imaginary part, isn't that right?

Reply:
I don't think so.
```python
v = np.linalg.eig([[1,2],[0,1]])
(v[0].dtype, v[1].dtype)
```
gives `(float, float)`. We could of course call `astype` at the end, otherwise `same_kind` won't work.

Reply:
Ok, I got it wrong. `safe` is more restrictive than `same_kind`. The documentation reads:
> - `equiv` means only byte-order changes are allowed.
> - `safe` means only casts which can preserve values are allowed.
> - `same_kind` means only safe casts or casts within a kind, like float64 to float32, are allowed.

Reply:
Numpy casts the result to float if the imaginary part is exactly zero: https://github.com/numpy/numpy/blob/master/numpy/linalg/linalg.py#L1325 I have no problem with undoing the cast.

Reply:
> I don't think so.
> 
> ```python
> v = np.linalg.eig([[1,2],[0,1]])
> (v[0].dtype, v[1].dtype)
> ```
> 
> gives `(float, float)`. We could of course call `astype` at the end, otherwise `same_kind` won't work.

I see.. well that is certainly something we cannot follow because we don't have variable dtypes. I suggest we add a boolean argument `real=True` to announce that eigenvalues will be real and add a runtime assertion to `evaluable.Eig`.

Reply:
> Ok, I got it wrong. `safe` is more restrictive than `same_kind`. The documentation reads:
> 
> > * `equiv` means only byte-order changes are allowed.
> > * `safe` means only casts which can preserve values are allowed.
> > * `same_kind` means only safe casts or casts within a kind, like float64 to float32, are allowed.
> 
> So, which one do we want?

Playing around with the different options only `equiv` forbids casting from int to float, but from the description I'm not sure that it is what we need. How about simply `asdtype(res.dtype) == instance.dtype`?

Reply:
Just to make things clearer, let's name the options `P` and `Q`:
- `P`: explicit cast in `Eig`
- `Q`: 
  1. test `dtype == dtype`
  2. test `can_cast(casting='safe')`
  3. test `can_cast(casting='same_kind')`, equivalent to `Q2 ∨ Q4`
  4. test the real 'same kind', i.e.`dtype.kind == dtype.kind`

So, which combination `c ∈ [P | ¬P] ✕ [Q1, Q2, Q3, Q4]` do you have in mind?

If either `Q2` or `Q3` is picked, even if we write out the cast, it's not checked/guaranteed. On the other hand, `Array.dtype ≥ value.dtype ≥ float` won't cause any problem at run time.

Reply:
I think there is no doubt we want Q1: the return type must equal what the object announced. The only question then is how to deal when the different kinds of integers and floats that we might come across, but I think `asdtype` covers it well.
You didn't define ¬P :)

Reply:
Btw I didn't see Joost's comment above, but a cast to complex seems fine to me as well. Either that or the extra flag I'd say, I don't see a third option.

Reply:
Ok, and that will imply `P`.

`¬P`: no explicit cast. 😃

Reply:
Let's go with that.

Reply:
Done. Thanks!